service/chain/follower.go

```diff
@@ -0,0 +1,316 @@
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/onflow/flow-go/follower"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/badger/operation"
+	"github.com/rs/zerolog"
+
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+type Follower struct {
+	log zerolog.Logger
+
+	consensus follower.ConsensusFollower
+	db        *badger.DB
+
+	availableHeight uint64
+	height          uint64
+	blockID         flow.Identifier
+}
+
+func FromFollower(log zerolog.Logger, follower follower.ConsensusFollower, db *badger.DB) *Follower {
+	f := Follower{
+		log: log,
+
+		consensus:       follower,
+		height:          math.MaxUint64,
+		availableHeight: math.MaxUint64,
+		blockID:         flow.ZeroID,
+	}
+
+	follower.AddOnBlockFinalizedConsumer(f.OnBlockFinalized)
+
+	return &f
+}
+
+func (f *Follower) OnBlockFinalized(finalizedBlockID flow.Identifier) {
+	var height uint64
+	err := f.db.View(operation.RetrieveFinalizedHeight(&height))
+	if err != nil {
+		f.log.Error().Err(err).Msg("Could not retrieve finalized block height")
+		return
+	}
+
+	f.availableHeight = height
+}
+
+func (f *Follower) Root() (uint64, error) {
+	var height uint64
+	err := f.db.View(operation.RetrieveRootHeight(&height))
+	if err != nil {
+		return 0, fmt.Errorf("could not look up root height: %w", err)
+	}
+	return height, nil
+}
+
+func (f *Follower) Commit(height uint64) (flow.StateCommitment, error) {
+
+	blockID, err := f.block(height)
+	if errors.Is(err, storage.ErrNotFound) {
+		return flow.StateCommitment{}, dps.ErrFinished
+	}
+	if errors.Is(err, dps.ErrTimeout) {
+		return flow.StateCommitment{}, dps.ErrTimeout
+	}
+	if err != nil {
+		return flow.StateCommitment{}, fmt.Errorf("could not get block for height: %w", err)
+	}
+
+	var commit flow.StateCommitment
+	err = f.db.View(operation.LookupStateCommitment(blockID, &commit))
+	if err != nil {
+		return flow.StateCommitment{}, fmt.Errorf("could not look up commit: %w", err)
+	}
+
+	return commit, nil
+}
+
+func (f *Follower) Header(height uint64) (*flow.Header, error) {
+
+	blockID, err := f.block(height)
+	if errors.Is(err, dps.ErrTimeout) {
+		return nil, dps.ErrTimeout
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not get block for height: %w", err)
+	}
+
+	var header flow.Header
+	err = f.db.View(operation.RetrieveHeader(blockID, &header))
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve header: %w", err)
+	}
+	return &header, nil
+}
+
+func (f *Follower) Collections(height uint64) ([]*flow.LightCollection, error) {
+
+	blockID, err := f.block(height)
+	if errors.Is(err, dps.ErrTimeout) {
+		return nil, dps.ErrTimeout
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not get block for height: %w", err)
+	}
+
+	var collIDs []flow.Identifier
+	err = f.db.View(operation.LookupPayloadGuarantees(blockID, &collIDs))
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup collections: %w", err)
+	}
```

Comment:
This might also not really work out. Thing is, we get certain things from the protocol state, and certain things from the execution node feed. Previously, we were working with the Badger DB of the execution node, which contains certain things such as collections. For the normal consensus follower, it doesn't really know collections, only guarantees. We thus need to have some kind of crossover between consensus follower and execution node feed.

I would say, create two components to consume the two streams, and then two other components to implement our inferfaces. The chain interfaces can then consume from both the follower and the execution feed.